demos/mqtt/subscription-manager/mqtt_subscription_manager.c

```diff
@@ -0,0 +1,448 @@
+ * use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies of
+ * the Software, and to permit persons to whom the Software is furnished to do so,
+ * subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in all
+ * copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS
+ * FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR
+ * COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER
+ * IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN
+ * CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
+ */
+
+/**
+ * @file mqtt_subscription_manager.c
+ * @brief Implementation of the API of a subscription manager for handling subscription callbacks
+ * to topic filters in MQTT operations.
+ */
+
+/* Standard includes. */
+#include <string.h>
+#include <assert.h>
+
+/* Include header for the subscription manager. */
+#include "mqtt_subscription_manager.h"
+
+/**
+ * @brief Represents a registered record of the topic filter and its associated callback
+ * in the subscription manager registry.
+ */
+typedef struct SubscriptionManager_Record
+{
+    const char * pTopicFilter;
+    uint16_t topicFilterLength;
+    SubscriptionManager_Callback_t callback;
+} SubscriptionManager_Record_t;
+
+/**
+ * @brief The default value for the maximum size of the callback registry in the
+ * subscription manager.
+ */
+#ifndef MAX_SUBSCRIPTION_CALLBACK_RECORDS
+    #define MAX_SUBSCRIPTION_CALLBACK_RECORDS    5
+#endif
+
+/**
+ * @brief The registry to store records of topic filters and their subscription callbacks.
+ */
+static SubscriptionManager_Record_t callbackRecordList[ MAX_SUBSCRIPTION_CALLBACK_RECORDS ] = { 0 };
+
+/**
+ * @brief Handle special corner cases regarding wildcards at the end of topic
+ * filters, as documented by the MQTT protocol spec.
+ *
+ * @param[in] pTopicFilter The topic filter containing the wildcard.
+ * @param[in] topicFilterLength Length of the topic filter being examined.
+ * @param[in] topicNameLength Length of the topic name being examined.
+ * @param[in] nameIndex Index of the topic name being examined.
+ * @param[in] filterIndex Index of the topic filter being examined.
+ * @param[out] pMatch Whether the topic filter and topic name match.
+ *
+ * @return Returns whether the topic filter and the topic name match.
+ */
+static bool matchEndWildcards( const char * pTopicFilter,
+                               uint16_t topicFilterLength,
+                               uint16_t topicNameLength,
+                               uint16_t nameIndex,
+                               uint16_t filterIndex );
+
+/**
+ * @brief Attempt to match characters in a topic filter by wildcards.
```

Comment:
Please elaborate the description. If the current character in the topic filter is '+', it consumes one level form the topic name. If the If the current character in the topic filter is '#', it assumes that the topic name matches and stops matching.

Reply:
Added description for the matching logic performed in the function